43558 light dark images (#43707)
mdn/content: files/en-us/web/css/reference/values/color_value/light-dark/index.md

Reviewer: - while here, can we move all this content to a "Description" section after "Syntax"?
- "are able to indicate" can be "can indicate"
- this content also needs to be updated to add `<image>` values
- shouldn't "user's preference is set" be "used color scheme is" to account for both user preference and any developer-set color scheme?

Author: done

## Syntax

```css-nolint
/* Named color values */
light-dark(
  black,
  white
);

/* RGB color values */
light-dark(
  rgb(0 0 0),
  rgb(255 255 255)
);

/* image url values */
light-dark(
  url("light-icon.png"),
  url("dark-icon.png")
);

/* linear-gradient values */
light-dark(
  linear-gradient(135deg, ghostwhite 20%, tomato),
  linear-gradient(45deg, darkslategray 20%, gold)
);

/* Custom properties */
light-dark(
  var(--light),
  var(--dark)
);
```

### Values

The light-dark() function has two forms:

- Accepts two `<color>` values:
  - `<color>` (light)
    - : The {{CSSXref("&lt;color&gt;")}} value used when the {{CSSXref("color-scheme")}} is light or no preference is set.
  - `<color>` (dark)
    - : The {{CSSXref("&lt;color&gt;")}} value used when the {{CSSXref("color-scheme")}} is dark`.
- Accepts two `<image>` values:
  - `<image>` (light)
    - : The {{CSSXref("&lt;image&gt;")}} value used when the {{CSSXref("color-scheme")}} is light or no preference is set.
  - `<image>` (dark)

>     - : The {{CSSXref("&lt;image&gt;")}} value used when the {{CSSXref("color-scheme")}} is dark`.

Reviewer: Backticks around `light` and `dark` need fixing

Author: done

  - `none`
    - : The `none` keyword produces a fully transparent image with no natural size.

## Description

> Users can indicate their color-scheme preference through their operating system settings (e.g., light or dark mode) or their user agent settings. The `light-dark()` function enables providing, either, two color values where any `<color>` value is accepted or two image values where any `<image>` value is accepted. The `light-dark()` CSS color function returns the first value if the used color scheme is to `light` or if no preference is set and the second value if the used color scheme is to `dark`.

Reviewer: ```suggestion
Users can indicate their color scheme preference through their operating system settings (e.g., light or dark mode) or their user agent settings. The `light-dark()` function enables providing either two color values where any `<color>` value is accepted or two image values where any `<image>` value is accepted. The `light-dark()` function returns the first value if the used color scheme is `light` or if no preference is set and the second value if the used color scheme is `dark`.
```

To enable support for the `light-dark()` color function, the {{CSSXref("color-scheme")}} must have a value of `light dark`, usually set on the {{CSSXref(":root")}} [pseudo-class](/en-US/docs/Web/CSS/Reference/Selectors/Pseudo-classes).

```css
:root {
  color-scheme: light dark;
}
body {
  color: light-dark(#333b3c, #efefec);
  background-color: light-dark(#efedea, #223a2c);
}
```

## Formal syntax

{{CSSSyntax}}

## Examples

> ### Setting colors based on color scheme

Reviewer: ```suggestion
### Setting colors based on a color scheme
```

sidebar: cssref
---

> The **`light-dark()`** [CSS](/en-US/docs/Web/CSS) [`<color>` function](/en-US/docs/Web/CSS/Reference/Values/Functions#color_functions) enables setting two colors or images for a property - returning one of the two color or image options by detecting if the developer has set a light or dark color scheme or the user has requested light or dark color theme - without needing to encase the theme colors within a [`prefers-color-scheme`](/en-US/docs/Web/CSS/Reference/At-rules/@media/prefers-color-scheme) [media feature](/en-US/docs/Web/CSS/Guides/Media_queries/Using#targeting_media_features) query.

Reviewer: - the intro can be shortened and tightened since the following paragraphs explain the details.
- suggesting to drop color as mentioned in https://github.com/mdn/content/pull/43708#discussion_r3065757432. LMK what you think.
```suggestion
The **`light-dark()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/Reference/Values/Functions) accepts two colors or two images and returns a color or an image based on the active color scheme, without needing a [`prefers-color-scheme`](/en-US/docs/Web/CSS/Reference/At-rules/@media/prefers-color-scheme) [media feature](/en-US/docs/Web/CSS/Guides/Media_queries/Using#targeting_media_features).
```

## Syntax

```css-nolint
/* Named color values */
color: light-dark(black, white);
background-color: light-dark(white, black);

/* RGB color values */
color: light-dark(
  rgb(0 0 0),
  rgb(255 255 255)
);

/* image url values */
color: light-dark(
  url("light-icon.png"),
  url("dark-icon.png")
);

/* linear-gradient values */
color: light-dark(
  linear-gradient(135deg, ghostwhite 20%, tomato),
  linear-gradient(45deg, darkslategray 20%, gold)

> );

Reviewer: Both these need properties that accept `<image>` values, so `background-image` or `list-style-image`, also shown in spec examples. There are some examples using linear-gradient in https://github.com/web-platform-tests/wpt/blob/master/css/css-color/light-dark-image.html.

But taking a step back, per our convention for the syntax block on CSS function pages, we should not show the full declaration, only the semantics of using the syntax. So we should drop the properties altogether from this block. See [template](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/CSS_function_page_template#syntax) and [color-mix() syntax block](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/color_value/color-mix#syntax.).
(I realize the syntax block on this page already had the property declarations, but since we're updating it, it's a good opportunity to fix it so it aligns with the template.)

Author: done

- `dark-color`
  - : {{CSSXref("&lt;color&gt;")}} value to be set for dark {{CSSXref("color-scheme")}}.

Functional notation: `light-dark(light-image, dark-image)`

- `light-image`
  - : {{CSSXref("&lt;image&gt;")}} value to be set for light {{CSSXref("color-scheme")}}.

- `dark-image`
  - : {{CSSXref("&lt;image&gt;")}} value to be set for dark {{CSSXref("color-scheme")}}.

> 

Reviewer: - The values here don't match terms in "Formal syntax".
- `none` keyword is missing from the image form list.
- Could you attempt a rewrite along these lines:

"The `light-dark()` function has two forms:
- Accepts two `<color>` values:
   - `<color>` (light)
     - : The {{CSSXref("&lt;color&gt;")}} value used when the {{CSSXref("color-scheme")}} is `light` or no preference is set.
   - `<color>` (dark)
     - : The {{CSSXref("&lt;color&gt;")}} value used when the `{{CSSXref("color-scheme")}} is `dark`.
- Accepts two `<image>` values:
   - `<image>`  (light)
     - : (desc)
   - `<image>`  (dark)
     - : (desc)
   - `none`
     - : (desc)

"

Author: done

{{EmbedLiveSample("setting_colors_based_on_color_scheme", "100%", 500)}}

> ### Setting images based on color scheme

Reviewer: Looking at the example code, this should probably be "Setting gradient background based on a color scheme"

Author: done

> This is using the same code as the previous example but includes a `<div>`, instead of a `<p>` of text, with a `background-image` property.

Reviewer: I guess you mean similar HTML code

```suggestion
This example uses the same HTML code as the previous example but includes a `<div>` instead of a `<p>`.
```

</p>
<div class="wrapper">
  <section>
    <h2>Automatic</h2>
    <div></div>
  </section>
  <section class="light">
    <h2>Light</h2>
    <div></div>
  </section>
  <section class="dark">
    <h2>Dark</h2>
    <div></div>
  </section>
</div>
```

#### CSS

```css-nolint hidden
:root {
  /* this has to be set to switch between light or dark */
  color-scheme: light dark;

  --light-bg: ghostwhite;
  --light-color: darkslategray;
  --light-code: tomato;

  --dark-bg: darkslategray;
  --dark-color: ghostwhite;
  --dark-code: gold;
}
* {
  background-color: light-dark(var(--light-bg), var(--dark-bg));
  color: light-dark(var(--light-color), var(--dark-color));
}
.wrapper {
  display: flex;
  justify-content: space-around;
  padding: 0.8rem;
}
.light {
  /* forces light color-scheme */
  color-scheme: light;
}
.dark {
  /* forces dark color-scheme */
  color-scheme: dark;
}
section {
  width: 25%;
  padding: 5px;
  color: light-dark(
    var(--light-code),
    var(--dark-code)
  );
  border: 2px solid light-dark(var(--light-code), var(--dark-code));
}
@supports (background-image: light-dark(url("light.png"), url("dark.png"))) {
  .supports {display:none;}
}
```

> Firstly, we define the light and dark `linear-gradient()`s using custom properties.

Reviewer: ```suggestion
Firstly, we define light and dark `linear-gradient()` values as custom properties.
```

```css
:root {
  /*  light dark gradients  */
  --light-grad: linear-gradient(135deg, var(--light-bg) 20%, var(--light-code));
  --dark-grad: linear-gradient(45deg, var(--dark-bg) 30%, var(--dark-code));
}
```

```css hidden
section div {
  width: 80%;
  aspect-ratio: 1/1;
  margin: auto;
  border: 1px solid light-dark(var(--light-code), var(--dark-code));
}
@supports not (
  background-image: light-dark(url("light.png"), url("dark.png"))
) {
  section div {
    width: 60%;
  }
}
```

> Then we use the custom properties to set the light and dark `background-image`.

Reviewer: ```suggestion
Then we use the custom properties with `light-dark()` to set the `background-image` property based on the active color scheme.
```

```css-nolint
section div {
  background-image: light-dark(
    var(--light-grad),
    var(--dark-grad)
  );
}
```

#### Result

> {{EmbedLiveSample("setting_colors_based_on_color_scheme", "100%", 350)}}

Reviewer: it will need to be updated to match the section heading

Author: done
